include/swift/AST/DiagnosticsSema.def

```diff
@@ -1530,6 +1530,9 @@ ERROR(types_not_equal,none,
       (Type, Type, Type))
 ERROR(type_does_not_conform_owner,none,
       "%0 requires that %1 conform to %2", (Type, Type, Type))
+ERROR(type_does_not_conform_decl_owner,none,
+      "%0 %1 requires that %2 conform to %3 [with %2 = %4]",
+      (DescriptiveDeclKind, DeclName, Type, Type, Type))
```

Comment:
Reading the output in the tests I find that `struct 'Decl1' requires that 'T' conform to 'P' [with 'T' = 'Y']` is a bit awkward, and I personally don't find referring to the generic type parameter ('T') upfront to be the easiest thing to parse, either.

How about this instead (for `Conforms<Y>.Decl1.self`):
`error: referencing 'Decl1' on 'Conforms<Y>' requires that 'Y' conform to 'P'`
`note: where T = Y` <--- pointing to the `struct Conforms<T>` declaration

?

Reply:
I think that's doable if I can detect that Y is an archetype, then I can refer to it's declaration instead, will try...

Reply:
Although I think we need to keep `requires that 'T' conform to 'P'` because there is not going to be a `Y` in the declaration error is pointing to.

Reply:
or maybe I should just re-do all fo the notes instead of including `[with ...]` into the diagnostic, just add it to the note like you suggested...